cpp/src/db/db_object.cuh

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cugraph.h>
+#include <vector>
+#include <map>
+#include "utilities/graph_utils.cuh"
+
+namespace cugraph {
+  /**
+   * Class for representing an entry in a pattern, which may either be a variable or constant value
+   */
+  template <typename idx_t>
+  class db_pattern_entry {
+    bool is_var;
+    idx_t constantValue;
+    std::string variableName;
+  public:
+    db_pattern_entry(std::string variable);
+    db_pattern_entry(idx_t constant);
+    db_pattern_entry(const db_pattern_entry<idx_t>& other);
+    bool isVariable() const;
+    idx_t getConstant() const;
+    std::string getVariable() const;
+  };
+
+  /**
+   * Class for representing a pattern (usually a triple pattern, but it's extensible)
+   */
+  template <typename idx_t>
+  class db_pattern {
+    std::vector<db_pattern_entry<idx_t>> entries;
+  public:
+    db_pattern();
+    db_pattern(const db_pattern<idx_t>& other);
+    int getSize() const;
+    const db_pattern_entry<idx_t>& getEntry(int position) const;
+    void addEntry(db_pattern_entry<idx_t>& entry);
+    bool isAllConstants();
+  };
+
```

Comment:
Before jumping to the more detailed review, I have a question, and how do this db_pattern & db_pattern_entry classes map to cypher queries?

Say for example, in the following query,

MATCH (:Person {name: 'Jennifer'})-[:WORKS_FOR]->(company:Company)
RETURN company

I guess this has three steps: 1) find the vertex with a label set to person and the name property set to 'Jennifer', 2) investigate the edges of the found vertex to find the edge with its label set to 'WORKS_FOR', and 3) traverse the edge to reach the company vertex.

This query is stored in a db_pattern object and each step becomes a db_pattern_entry? Or am I mis-guessing? Any corrections?

Reply:
You are mostly right. In your query we would first find all vertices with label:Person, then from those vertices find ones that have name:'Jennifer', then from those vertices find edges with type:WORKS_FOR, then from the ending vertices of the last find vertices with label:Company and return them. Each step of the query would be represented by a db_pattern object.

Reply:
Got it, so each step will be mapped to a db_pattern object (not a db_pattern_entry). I think it will be greatly helpful to add comments explaining this mapping in code readability.